libcontainer/dmz/_dmz.c

```diff
@@ -11,5 +13,17 @@ int main(int argc, char **argv)
 {
 	if (argc < 1)
 		return 127;
-	return execve(argv[0], argv, environ);
+	int r = execve(argv[0], argv, environ);
+	if (r) {
+		char *prefix = "exec ";
+		int err_len = strlen(prefix) + strlen(argv[0]) + 1;
+		char err[err_len];
+
+		strcpy(err, prefix);
+		strcpy(err + strlen(prefix), argv[0]);
+		err[err_len - 1] = '\0';
+
+		perror(err);
```

Comment:
Wouldn't something like this be simpler?

```suggestion
		char err[5 + PATH_MAX] = "exec "; // "exec " + argv[0]
		strlcat(err, argv[0], sizeof(err));
		err[sizeof(err) - 1] = '\0';
		perror(err);
```

No dynamic memory management, no VFAs, and no error-prone pointer maths.

Unfortunately, for some reason `strlcat` has issues even though it is defined in `nolibc` (as is `strlen`)...


```
make[1]: Entering directory '/home/cyphar/src/github.com/opencontainers/runc/libcontainer/dmz'
gcc -fno-asynchronous-unwind-tables -fno-ident -s -Os -nostdlib -lgcc -static -o binary/runc-dmz _dmz.c
/usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: /tmp/ccIfhSHD.o: in function `main':
_dmz.c:(.text.startup+0x5e): undefined reference to `strlen'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:18: binary/runc-dmz] Error 1
```

Reply:
Yes, weird. This might fun to investigate. Just removing the `__attribute__((unused))` from the function definition in nolibc, makes it work, though. So, for some reason it is removing a function that is used at link-time.

But for our use case, we can just do that with `strcpy()` and be happy. So I've updated it to that :)

Reply:
Heh, with `strncat()` it's simpler, updated to that :)

Reply:
btw, I've chased down the bug to use `strlcpy()` and fixed it upstream, the series was applied a few months ago: https://lore.kernel.org/all/20240219204821.GA9819@1wt.eu/

The issue was gcc being too smart and replacing things that implement strlen/friends with a call to its builtin implementation, that when we compile without stdlib it of course fails.

Needless to say, we don't need to do anything in runc, as we are not using `strlcpy()`.